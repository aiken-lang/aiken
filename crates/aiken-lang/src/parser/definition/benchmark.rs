use crate::{
    ast,
    parser::{error::ParseError, token::Token},
};
use chumsky::prelude::*;

pub fn parser() -> impl Parser<Token, ast::UntypedDefinition, Error = ParseError> {
    crate::parser::definition::test_like::parser(Token::Benchmark)
}

#[cfg(test)]
mod tests {
    use crate::assert_definition;

    #[test]
    fn def_benchmark() {
        assert_definition!(
            r#"
            bench foo(x via fuzz.any_int) {
                True
            }
            "#
        );
    }

    #[test]
    fn def_invalid_benchmark() {
        assert_definition!(
            r#"
            bench foo(x via f, y via g) {
                True
            }
            "#
        );
    }

    #[test]
    fn def_benchmark_annotated_fuzzer() {
        assert_definition!(
            r#"
            bench foo(x: Int via foo()) {
                True
            }
            "#
        );
    }
}
