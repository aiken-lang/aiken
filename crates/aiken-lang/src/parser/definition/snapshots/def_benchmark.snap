---
source: crates/aiken-lang/src/parser/definition/benchmark.rs
assertion_line: 136
description: "Code:\n\nbench foo(x via fuzz.any_int) {\n    True\n}\n"
snapshot_kind: text
---
Benchmark(
    Function {
        arguments: [
            ArgVia {
                arg: UntypedArg {
                    by: ByName(
                        Named {
                            name: "x",
                            label: "x",
                            location: 10..11,
                        },
                    ),
                    location: 10..11,
                    annotation: None,
                    doc: None,
                    is_validator_param: false,
                },
                via: FieldAccess {
                    location: 16..28,
                    label: "any_int",
                    container: Var {
                        location: 16..20,
                        name: "fuzz",
                    },
                },
            },
        ],
        body: Var {
            location: 36..40,
            name: "True",
        },
        doc: None,
        location: 0..29,
        name: "foo",
        public: false,
        return_annotation: None,
        return_type: (),
        end_position: 41,
        on_test_failure: FailImmediately,
    },
)
